tendermint/src/types/proto/crypto.rs

```diff
@@ -1,21 +1,53 @@
-#[derive(Clone, PartialEq, Eq, ::prost::Message, serde::Serialize, serde::Deserialize)]
-pub struct PublicKey {
-    #[prost(oneof = "public_key::Sum", tags = "1, 2")]
-    pub sum: Option<public_key::Sum>,
+use tracing::error;
+
+#[derive(Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type", content = "value")]
+pub enum PublicKey {
+    #[serde(
+        rename = "tendermint/PubKeyEd25519",
+        with = "crate::types::serializers::bytes::base64string"
+    )]
+    Ed25519(Vec<u8>), //TODO: should we check that bytes contain a valid public key?
+    #[serde(
+        rename = "tendermint/PubKeySecp256k1",
+        with = "crate::types::serializers::bytes::base64string"
+    )]
+    Secp256k1(Vec<u8>), //TODO: should we check that bytes contain a valid public key?
+}
+
+impl PublicKey {
+    pub fn raw(&self) -> Vec<u8> {
+        match self {
+            PublicKey::Ed25519(value) => value.clone(),
+            PublicKey::Secp256k1(value) => value.clone(),
+        }
+    }
 }
 
 impl From<PublicKey> for inner::PublicKey {
-    fn from(PublicKey { sum }: PublicKey) -> Self {
-        Self {
-            sum: sum.map(Into::into),
+    fn from(key: PublicKey) -> Self {
+        match key {
+            PublicKey::Ed25519(value) => inner::PublicKey {
+                sum: Some(inner::Sum::Ed25519(value)),
+            },
+            PublicKey::Secp256k1(value) => inner::PublicKey {
+                sum: Some(inner::Sum::Secp256k1(value)),
+            },
         }
     }
 }
 
 impl From<inner::PublicKey> for PublicKey {
     fn from(inner::PublicKey { sum }: inner::PublicKey) -> Self {
-        Self {
-            sum: sum.map(Into::into),
+        let Some(sum) = sum else {
```

Comment:
Well... I won't agree with you on this. This is unexpected behavior for `From` implementation. Better use method name `parse_or_terminate` or handle this logic on upper levels.

Reply:
I agree... I've pushed a change